cql3/query_processor.cc

```diff
@@ -351,7 +351,7 @@ query_processor::query_processor(service::storage_proxy& proxy, database& db, se
                     sm::make_derive(
                             "unprivileged_entries_evictions_on_size",
                             [] { return prepared_statements_cache::shard_stats().unprivileged_entries_evictions_on_size; },
-                            sm::description("Counts a number of prepared statements evictions from the unprivileged cache section due to a cache size restrictions. This is usually a sign of a cache pollution.")),
+                            sm::description("Counts a number of evictions of prepared statements from the prepared statements cache after they have been used only once. If this counter gets incremented this is a red flag.")),
```

Comment:
A "red flag" for what? Why?
If I understand correctly, it means the user might be preparing a new statement on each request instead of reusing a previously prepared statement?

Reply:
Yes, and if he was preparing the same query exactly (e.g. "SELECT * from table where key = ?", then it would not cause a cache eviction but rather a "cache hit" since prepared statements are cached by a query string name.

However if queries are being evicted right after they were used only once means that they are actually used only once, e.g. as described in #8674. 

And this is exactly the opposite from what's the intention of prepared statements and hence it's a "red flag".

Reply:
@vladzcloudius I need to make something clear about my review - and all my reviews in general.
When I say that something is not clear, it doesn't always mean I want an explanation now. It often means that I thought that the code or error message won't be clear to future users or reader as well - not just to me. In this case I noted that a message like "this is a red flag" sounds scary but doesn't help understand what's wrong. If the problem is that the user is probably preparing different statements and then using them only once - why not say exactly this? E.g.,

"Counts a number of evictions of prepared statements from the prepared statements cache after they have been used only once. An increasing counter suggests the user may be preparing a different statement for each request instead of reusing the same prepared statement with parameters".

By the way, the other case where the user is preparing exactly the same statement over and over with exactly the same string, do we have a counter for that too? Isn't this a problem too?

Reply:
> By the way, the other case where the user is preparing exactly the same statement over and over with exactly the same string, do we have a counter for that too? Isn't this a problem too?

@nyh It's not really an issue in a general case because it could be happening simply because the user has a worker that uses the same (good) parametrized query but the worker is short lived, so instances of that worker are being created over and over again and then they finish and exit. 

Another case similar to the above is when multiple instances of that worker are running concurrently and the first thing a single worker does is calling a PREPARE.

We already have this patch very optimized. In that case no CQL parsing is expected for the second time PREPARE is called.